core/lib/node/src/task/mod.rs

```diff
@@ -0,0 +1,76 @@
+//! Tasks define the "runnable" concept of the node, e.g. something that can be launched and runs until the node
+//! is stopped.
+//!
+//! Task is normally defined by implementing two traits [`IntoZkSyncTask`], which acts like a constructor, and
+//! [`ZkSyncTask`], which provides an interface for `ZkSyncNode` to implement the task lifecycle.
+
+use futures::future::BoxFuture;
+use zksync_health_check::CheckHealth;
+
+use crate::node::{NodeContext, StopReceiver};
+
+pub mod healtcheck_server;
+pub mod metadata_calculator;
+pub mod prometheus_exporter;
+
+/// Factory that can create a task.
+// Note: This have to be a separate trait, since `ZkSyncTask` has to be object-safe.
+pub trait IntoZkSyncTask: 'static + Send + Sync {
+    /// Config type for the task.
+    /// It is highly recommended for tasks to have dedicated configs, not shared with other tasks and resources.
+    type Config: 'static + Send + Sync;
+
+    /// Creates a new task.
+    /// `NodeContext` provides an interface to the utilities that task may need, e.g. the runtime handle to perform asynchronous
+    /// calls, or access to the resources.
+    fn create(
+        node: &NodeContext<'_>,
+        config: Self::Config,
+    ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>;
+}
+
+/// A task implementation.
+#[async_trait::async_trait]
+pub trait ZkSyncTask: 'static + Send + Sync {
+    /// Gets the healthcheck for the task, if it exists.
+    /// Guaranteed to be called only once per task.
+    fn healthcheck(&mut self) -> Option<Box<dyn CheckHealth>>;
+
+    /// Runs the task.
+    ///
+    /// Once any of the task returns, the node will shutdown.
+    /// If the task returns an error, the node will spawn an error-level log message and will return a non-zero
+    /// exit code.
+    ///
+    /// `stop_receiver` argument contains a channel receiver that will change its value once the node requests
+    /// a shutdown. Every task is expected to either await or periodically check the state of channel and stop
+    /// its execution once the channel is changed.
+    ///
+    /// Each task is expected to perform the required cleanup after receiving the stop signal (e.g. make sure that
+    /// spawned sub-tasks are awaited).
+    async fn run(self: Box<Self>, stop_receiver: StopReceiver) -> anyhow::Result<()>;
+
+    /// Asynchronous hook that will be called after *each task* has finished their cleanup.
+    /// It is guaranteed that no other task is running at this point, e.g. `ZkSyncNode` will invoke
+    /// this hook sequentially for each task.
```

Comment:
This hook is for stuff like automatic rollbacks if reorg detector needs them. We can't do that at the end of `run`, since some tasks may still be running when we'll start to rollback the database, and we can't really do that after `ZkSyncNode::run()`, since 1) `ZkSyncNode` owns the runtime that was used to initialize async stuff, and passing things through runtimes tends to cause problems, and 2) in the current form the task defines whether it needs to do something (e.g. it's a part of task logic), and if it's done outside of `ZkSyncNode`, it becomes somewhat arbitraty (e.g. `if task_was_resent { do_task_related_logic() } `).